### DISCUSSION TITLE
General README file, suggest that consider keeping things closer to llama.cpp for now?

### DISCUSSION THREAD

**Discussion by linuxmagic-mp** [cmp-nct/ggllm.cpp#21] [General]
For the record, the cmake breaks on my Linux machine, be nice if the 'make' instructions also included in the README.  For the record, there are more warnings in this repo, while performing the make..

**Comment by linuxmagic-mp** [comment]
Ooops, fat fingers committed comment to fast.  Thought I would like the make warnings here for cleanup.. 

ggml.c:4215:75: warning: passing argument 3 of ‘ggml_sizeof_tensor_data_impl’ from incompatible pointer type [-Wincompatible-pointer-types]
 4215 |     return ggml_tensor_overhead() + ggml_sizeof_tensor_data_impl(type, 2, &ne);
      |                                                                           ^~~
      |                                                                           |
      |                                                                           const int64_t (*)[2] {aka const long int (*)[2]}
ggml.c:4198:85: note: expected ‘const int64_t *’ {aka ‘const long int *’} but argument is of type ‘const int64_t (*)[2]’ {aka ‘const long int (*)[2]’}
 4198 | size_t ggml_sizeof_tensor_data_impl(enum ggml_type type, int n_dims, const int64_t* ne) {
      |                                                                      ~~~~~~~~~~~~~~~^~
ggml.c: In function ‘ggml_sizeof_tensor_3d’:
ggml.c:4220:75: warning: passing argument 3 of ‘ggml_sizeof_tensor_data_impl’ from incompatible pointer type [-Wincompatible-pointer-types]
 4220 |     return ggml_tensor_overhead() + ggml_sizeof_tensor_data_impl(type, 3, &ne);
      |                                                                           ^~~
      |                                                                           |
      |                                                                           const int64_t (*)[3] {aka const long int (*)[3]}
ggml.c:4198:85: note: expected ‘const int64_t *’ {aka ‘const long int *’} but argument is of type ‘const int64_t (*)[3]’ {aka ‘const long int (*)[3]’}
 4198 | size_t ggml_sizeof_tensor_data_impl(enum ggml_type type, int n_dims, const int64_t* ne) {
      |                                                                      ~~~~~~~~~~~~~~~^~
ggml.c: In function ‘ggml_sizeof_tensor_4d’:
ggml.c:4225:75: warning: passing argument 3 of ‘ggml_sizeof_tensor_data_impl’ from incompatible pointer type [-Wincompatible-pointer-types]
 4225 |     return ggml_tensor_overhead() + ggml_sizeof_tensor_data_impl(type, 4, &ne);
      |                                                                           ^~~
      |                                                                           |
      |                                                                           const int64_t (*)[4] {aka const long int (*)[4]}
ggml.c:4198:85: note: expected ‘const int64_t *’ {aka ‘const long int *’} but argument is of type ‘const int64_t (*)[4]’ {aka ‘const long int (*)[4]’}
 4198 | size_t ggml_sizeof_tensor_data_impl(enum ggml_type type, int n_dims, const int64_t* ne) {
      |                                                                      ~~~~~~~~~~~~~~~^~
ggml.c: In function ‘graph_calculate_mean_sd’:
ggml.c:17043:18: warning: implicit conversion from ‘float’ to ‘double’ to match other operand of binary expression [-Wdouble-promotion]
17043 |         variance += pow(array[i] - *mean, 2);
      |                  ^~
examples/train-text-from-scratch/train-text-from-scratch.cpp:2371:21: warning: suggest parentheses around ‘-’ in operand of ‘&’ [-Wparentheses]
 2371 |         file->seek(0-file->tell() & 31, SEEK_CUR);
      |                    ~^~~~~~~~~~~~~
examples/train-text-from-scratch/train-text-from-scratch.cpp:2386:17: warning: suggest parentheses around ‘-’ in operand of ‘&’ [-Wparentheses]
 2386 |     file->seek(0-file->tell() & 31, SEEK_CUR);
      |                ~^~~~~~~~~~~~~
examples/train-text-from-scratch/train-text-from-scratch.cpp: In function ‘void read_tensor(llama_file*, ggml_tensor*)’:
examples/train-text-from-scratch/train-text-from-scratch.cpp:2407:17: warning: suggest parentheses around ‘-’ in operand of ‘&’ [-Wparentheses]
 2407 |     file->seek(0-file->tell() & 31, SEEK_CUR);
      |                ~^~~~~~~~~~~~~
In file included from /usr/include/string.h:495,
                 from /usr/include/c++/9/cstring:42,
                 from examples/train-text-from-scratch/train-text-from-scratch.cpp:7:
In function ‘char* strncpy(char*, const char*, size_t)’,
    inlined from ‘void init_model(my_llama_model*)’ at examples/train-text-from-scratch/train-text-from-scratch.cpp:305:16:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:34: warning: ‘char* __builtin_strncpy(char*, const char*, long unsigned int)’ specified bound 64 equals destination size [-Wstringop-truncation]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘char* strncpy(char*, const char*, size_t)’,
    inlined from ‘void init_model(my_llama_model*)’ at examples/train-text-from-scratch/train-text-from-scratch.cpp:306:16:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:34: warning: ‘char* __builtin_strncpy(char*, const char*, long unsigned int)’ specified bound 64 equals destination size [-Wstringop-truncation]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
In function ‘char* strncpy(char*, const char*, size_t)’,
    inlined from ‘void init_model(my_llama_model*)’ at examples/train-text-from-scratch/train-text-from-scratch.cpp:307:16:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:34: warning: ‘char* __builtin_strncpy(char*, const char*, long unsigned int)’ specified bound 64 equals destination size [-Wstringop-truncation]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by cmp-nct** [comment]
I pushed a maintenance update today which solved a lot of the warnings.
The random problems with cmake and make have been solved, at least I hope so.
Readme is new and contains good instructions now :)

**Reply by linuxmagic-mp** [reply]
Better..

One left..

```
libfalcon.cpp: In member function ‘void falcon_file_loader::read_vocab()’:
libfalcon.cpp:515:32: warning: comparison of integer expressions of different signedness: ‘uint32_t’ {aka ‘unsigned int’} and ‘int32_t’ {aka ‘int’} [-Wsign-compare]
  515 |         for (uint32_t i = 0; i < hparams.n_vocab; i++) {
      |                              ~~^~~~~~~~~~~~~~~~~
libfalcon.cpp: In function ‘bool falcon_eval_internal(falcon_context&, const llama_token*, int, int, int, const char*, int)’:
libfalcon.cpp:1549:20: warning: variable ‘offload_func_nr’ set but not used [-Wunused-but-set-variable]
 1549 |     offload_func_t offload_func_nr = llama_nop; // nr = non-repeating
      |                    ^~~~~~~~~~~~~~~
libfalcon.cpp:1550:20: warning: variable ‘offload_func_kqv’ set but not used [-Wunused-but-set-variable]
 1550 |     offload_func_t offload_func_kqv = llama_nop;
      |                    ^~~~~~~~~~~~~~~~
libfalcon.cpp:1788:20: warning: variable ‘offload_func’ set but not used [-Wunused-but-set-variable]
 1788 |     offload_func_t offload_func = llama_nop;
      |                    ^~~~~~~~~~~~
```

**Reply by cmp-nct** [reply]
That one should be fixed too, the offload will be used in a while.
Though there are hundreds of warnings on conversion, they are all suppressed
